### PR TITLE
fix for PHP 8

### PR DESCRIPTION
**Pull request by remicollet**
Also allow old phpunit for old PHP versions

### CODE REVIEW

File: lib/QrReader.php [hunk 1]
```diff
@@ -62,7 +62,7 @@ public function __construct($imgSource, $sourceType = QrReader::SOURCE_TYPE_FILE
             $height = $im->getImageHeight();
             $source = new IMagickLuminanceSource($im, $width, $height);
         } else {
-            if (!is_resource($im)) {
+            if (!is_resource($im) && !is_object($im)) {
```
Comment (remicollet) [inline lib/QrReader.php:65]: Resource to object change in PHP 8

File: lib/GDLuminanceSource.php [hunk 2]
```diff
@@ -65,8 +65,8 @@ public function GDLuminanceSource($gdImage, $width, $height)
 
         for ($j = 0; $j < $height; $j++) {
             for ($i = 0; $i < $width; $i++) {
-                $argb  = imagecolorat($this->$gdImage, $i, $j);
-                $pixel = imagecolorsforindex($this->$gdImage, $argb);
+                $argb  = imagecolorat($this->gdImage, $i, $j);
+                $pixel = imagecolorsforindex($this->gdImage, $argb);
```
Comment (remicollet) [inline lib/GDLuminanceSource.php:69]: Previous have probably never worked....

**Comment by remicollet** [comment]
```
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 5.6.40
Configuration: /dev/shm/BUILD/php-qrcode-detector-decoder-07fceefb79d895e858e52921afb9c1433d2f3d5e/phpunit.xml.dist

.                                                                   1 / 1 (100%)

Time: 60 ms, Memory: 26.00MB

OK (1 test, 1 assertion)


PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.0.33
Configuration: /dev/shm/BUILD/php-qrcode-detector-decoder-07fceefb79d895e858e52921afb9c1433d2f3d5e/phpunit.xml.dist

.                                                                   1 / 1 (100%)

Time: 40 ms, Memory: 6.00MB

OK (1 test, 1 assertion)


PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.33
Configuration: /dev/shm/BUILD/php-qrcode-detector-decoder-07fceefb79d895e858e52921afb9c1433d2f3d5e/phpunit.xml.dist

  Warning - The configuration file did not pass validation!
  The following problems have been detected:

  Line 3:
  - Element 'coverage': This element is not expected.

  Test results may not be as expected.


.                                                                   1 / 1 (100%)

Time: 30 ms, Memory: 10.01 MB

OK (1 test, 1 assertion)


PHPUnit 8.5.11 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.34
Configuration: /dev/shm/BUILD/php-qrcode-detector-decoder-07fceefb79d895e858e52921afb9c1433d2f3d5e/phpunit.xml.dist

  Warning - The configuration file did not pass validation!
  The following problems have been detected:

  Line 3:
  - Element 'coverage': This element is not expected.

  Test results may not be as expected.


.                                                                   1 / 1 (100%)

Time: 31 ms, Memory: 10.01 MB

OK (1 test, 1 assertion)



Runtime:       PHP 7.3.25
Configuration: /dev/shm/BUILD/php-qrcode-detector-decoder-07fceefb79d895e858e52921afb9c1433d2f3d5e/phpunit.xml.dist

.                                                                   1 / 1 (100%)

Time: 00:00.018, Memory: 10.01 MB

OK (1 test, 1 assertion)


PHPUnit 9.4.3 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.13
Configuration: /dev/shm/BUILD/php-qrcode-detector-decoder-07fceefb79d895e858e52921afb9c1433d2f3d5e/phpunit.xml.dist

.                                                                   1 / 1 (100%)

Time: 00:00.025, Memory: 10.01 MB

OK (1 test, 1 assertion)


PHPUnit 9.4.3 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.0.0
Configuration: /dev/shm/BUILD/php-qrcode-detector-decoder-07fceefb79d895e858e52921afb9c1433d2f3d5e/phpunit.xml.dist

.                                                                   1 / 1 (100%)

Time: 00:00.027, Memory: 6.00 MB

OK (1 test, 1 assertion)

```

**Comment by francislavoie** [comment]
Thanks @remicollet 😄 

FYI @khanamiryan this is a blocker for https://github.com/endroid/qr-code/issues/282 and by proxy https://github.com/Vectorface/GoogleAuthenticator/pull/2. It would be great if you could merge this soon!

File: lib/QrReader.php (hunk 1 above)
Comment (francislavoie) [reply lib/QrReader.php:65]: I think it should be done like this actually, for proper type handling

```suggestion
            if ($im === false) {
```

PHPStan would complain: https://phpstan.org/r/15d5bc46-209a-4f2c-bc01-5e29daddb8fc

With `=== false`: https://phpstan.org/r/e1632f88-9ea3-45ef-84e6-86cbbaddc649

Comment (remicollet) [reply lib/QrReader.php:65]: Sorry, but I really think phpstan this warning have no value here.

Comment (francislavoie) [reply lib/QrReader.php:65]: I'm not stating PHPStan as the reason, but rather as an example. This is a simpler check that actually targets the correct type returned in both PHP 7 and 8. I don't see a reason not to simplify it here.

**Comment by kohlerdominik** [comment]
Is there anything blocking this from merge @khanamiryan ?

**Comment by volkv** [comment]
@khanamiryan up

**Comment by strebl** [comment]
hi @khanamiryan! Can we help you to speed up the merge process?

**Comment by khanamiryan** [comment]
Sorry, I didn't receive notifications. I don't know why.

**Comment by endroid** [comment]
Thank you @khanamiryan!

**Pull request merged**